File: python/652.py
"""Project Euler Problem 652: Distinct values of logarithms.

Find the number of distinct values of log_m(n) for 2 ≤ m,n ≤ N.

There are (N - 1)² ways to choose m and n, but there are duplicates that we
need to subtract:
- log_m(n) = 1 if m = n
- If m and n are both k-th powers, then log_m(n) = log_{ᵏ√m}(ᵏ√n)
- If n is a k-th power of m ≥ 3, then log_m(n) is identical to the value
  when m = 2 and n = 2^k
- Pairs of powers with exponents that are relatively prime
"""

from __future__ import annotations

from math import gcd, isqrt

from sympy import mobius, primerange


def mod_inverse(a: int, m: int) -> int:
    """Modular inverse using extended Euclidean algorithm."""
    if m == 1:
        return 0
    t, new_t = 0, 1
    r, new_r = m, a % m
    while new_r != 0:
        q = r // new_r
        t, new_t = new_t, t - q * new_t
        r, new_r = new_r, r - q * new_r
    if r != 1:
        raise ValueError("Modular inverse does not exist")
    if t < 0:
        t += m
    return t


def pow_mod(base: int, exp: int, mod: int) -> int:
    """Modular exponentiation."""
    result = 1
    base %= mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def nth_root(n: int, k: int) -> int:
    """Compute k-th root of n (floor)."""
    if k == 1:
        return n
    if k == 2:
        return isqrt(n)
    low, high = 1, n
    while low < high:
        mid = (low + high + 1) // 2
        power = pow(mid, k)
        if power <= n:
            low = mid
        else:
            high = mid - 1
    return low


def nCr(n: int, k: int) -> int:
    """Binomial coefficient."""
    if k < 0 or k > n:
        return 0
    if k == 0 or k == n:
        return 1
    result = 1
    for i in range(min(k, n - k)):
        result = result * (n - i) // (i + 1)
    return result


def pre_mobius(max_k: int) -> list[int]:
    """Precompute Möbius function."""
    mu = [1] * (max_k + 1)
    mu[0] = 0
    primes = list(primerange(2, max_k + 1))
    for p in primes:
        for j in range(p, max_k + 1, p):
            mu[j] = -mu[j]
        for j in range(p * p, max_k + 1, p * p):
            mu[j] = 0
    return mu


def solve() -> int:
    """Solve Problem 652."""
    N = 10**18
    M = 10**9

    max_k = 60  # log_3(10^18) ≈ 37
    mobius_vals = pre_mobius(max_k)

    ans = ((N - 1) ** 2 - (N - 2)) % M

    # Subtract k-th powers
    for k in range(2, max_k + 1):
        if pow(3, k) > N:
            break
        num_powers = nth_root(N, k) - 1
        ans = (ans + 2 * mobius_vals[k] * nCr(num_powers, 2)) % M

    # Subtract cases where n is a k-th power of m ≥ 3
    for k in range(2, max_k + 1):
        if pow(3, k) > N:
            break
        ans = (ans - 2 * (nth_root(N, k) - 2)) % M
        e = 2
        while pow(2, k * e) <= N:
            ans = (ans - 2 * mobius_vals[e] * (nth_root(N, k * e) - 1)) % M
            e += 1

    # Subtract pairs of powers with relatively prime exponents
    max_b = int(N ** (1 / 3)) + 1
    is_perfect_power = [False] * (max_b + 1)
    for b in range(2, max_b + 1):
        if pow(b, 6) > N:
            break
        e = 2
        while pow(b, 3 * e) <= N:
            power_val = pow(b, e)
            if power_val <= max_b:
                is_perfect_power[power_val] = True
            e += 1

    for b in range(3, max_b + 1):
        if pow(b, 3) > N:
            break
        if is_perfect_power[b]:
            continue
        largest_e = 1
        while pow(b, largest_e + 1) <= N:
            largest_e += 1
        for e1 in range(2, largest_e + 1):
            for e2 in range(2, largest_e + 1):
                if gcd(e1, e2) == 1:
                    ans = (ans - 1) % M

    return ans % M


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/652.c
/*
 * Project Euler 652 - Distinct values of logarithms
 * Count distinct log_m(n) for 2 <= m,n <= N = 10^18, mod 10^9.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef long long ll;
typedef __int128 lll;

#define M 1000000000LL

static ll nth_root(ll n, int k) {
    if (k == 1) return n;
    if (n <= 1) return n;
    double dr = 1.0 / k;
    ll x = (ll)pow((double)n, dr);
    /* Refine */
    if (x < 1) x = 1;
    /* Check x-1, x, x+1, x+2 */
    for (ll t = x + 2; t >= 1 && t >= x - 2; t--) {
        /* Check t^k <= n */
        lll pw = 1;
        int ok = 1;
        for (int i = 0; i < k; i++) {
            pw *= t;
            if (pw > n) { ok = 0; break; }
        }
        if (ok) {
            /* Check (t+1)^k > n */
            lll pw2 = 1;
            int over = 0;
            for (int i = 0; i < k; i++) {
                pw2 *= (t + 1);
                if (pw2 > n) { over = 1; break; }
            }
            if (over) return t;
        }
    }
    return x;
}

/* Mobius function precomputed */
static int mobius_vals[65];

static void precompute_mobius(int max_k) {
    for (int i = 0; i <= max_k; i++) mobius_vals[i] = 1;
    mobius_vals[0] = 0;

    /* Sieve of primes up to max_k */
    int is_prime[65];
    for (int i = 0; i <= max_k; i++) is_prime[i] = 1;
    is_prime[0] = is_prime[1] = 0;
    for (int i = 2; i * i <= max_k; i++)
        if (is_prime[i])
            for (int j = i*i; j <= max_k; j += i)
                is_prime[j] = 0;

    for (int p = 2; p <= max_k; p++) {
        if (!is_prime[p]) continue;
        for (int j = p; j <= max_k; j += p)
            mobius_vals[j] = -mobius_vals[j];
        for (int j = p*p; j <= max_k; j += p*p)
            mobius_vals[j] = 0;
    }
}

static ll nCr2(ll n) {
    /* C(n, 2) = n*(n-1)/2 */
    return (n % M) * ((n - 1) % M) % M * 500000000LL % M; /* inv2 mod 10^9 doesn't work since 10^9 is not prime */
}

/* Since M = 10^9 is not prime, compute C(n,2) mod M directly */
static ll nCr2_mod(ll n) {
    /* n*(n-1)/2 mod M */
    ll a = n % (2 * M);
    ll b = (n - 1) % (2 * M);
    ll prod = (a * b) % (2 * M);
    return (prod / 2) % M;
}

/* Check if b^e fits */
static int pow_fits(ll b, int e, ll limit) {
    lll pw = 1;
    for (int i = 0; i < e; i++) {
        pw *= b;
        if (pw > limit) return 0;
    }
    return pw <= limit;
}

int main(void) {
    ll N = 1000000000000000000LL; /* 10^18 */

    precompute_mobius(60);

    /* ans = (N-1)^2 - (N-2) mod M */
    ll ans = ((N - 1) % M * ((N - 1) % M) % M - (N - 2) % M + M) % M;

    /* Subtract k-th powers */
    for (int k = 2; k <= 60; k++) {
        if (!pow_fits(3, k, N)) break;
        ll num_powers = nth_root(N, k) - 1;
        ll contrib = (2 * mobius_vals[k] % M + M) % M * nCr2_mod(num_powers) % M;
        ans = (ans + contrib) % M;
    }

    /* Subtract cases where n is a k-th power of m >= 3 */
    for (int k = 2; k <= 60; k++) {
        if (!pow_fits(3, k, N)) break;
        ans = (ans - 2 * ((nth_root(N, k) - 2) % M) % M + 2 * M) % M;
        int e = 2;
        while (pow_fits(2, k * e, N)) {
            ll val = (2 * (ll)mobius_vals[e] % M + M) % M * ((nth_root(N, k * e) - 1) % M) % M;
            ans = (ans - val % M + M) % M;
            e++;
        }
    }

    /* Subtract pairs of powers with relatively prime exponents */
    /* For each base b >= 3 that is not a perfect power */
    ll max_b_val = nth_root(N, 3);

    /* Mark perfect powers up to max_b_val */
    /* A number is a perfect power if it can be written as c^e for e >= 2 */
    /* We only need to check for small max_b_val */
    /* Since max_b_val ~ 10^6, we can use an array */
    int max_b = (int)max_b_val;
    if (max_b > 10000000) max_b = 10000000;
    char *is_perfect_power = (char *)calloc(max_b + 1, 1);

    for (int b = 2; b <= max_b; b++) {
        if (!pow_fits(b, 6, N)) break;
        int e = 2;
        while (1) {
            lll pw = 1;
            int ok = 1;
            for (int i = 0; i < 3 * e; i++) {
                pw *= b;
                if (pw > N) { ok = 0; break; }
            }
            if (!ok) break;
            /* b^e is a perfect power */
            lll bpow = 1;
            for (int i = 0; i < e; i++) bpow *= b;
            if (bpow <= max_b) is_perfect_power[(int)bpow] = 1;
            e++;
        }
    }

    for (int b = 3; b <= max_b; b++) {
        if (!pow_fits(b, 3, N)) break;
        if (is_perfect_power[b]) continue;
        int largest_e = 1;
        while (pow_fits(b, largest_e + 1, N)) largest_e++;
        for (int e1 = 2; e1 <= largest_e; e1++) {
            for (int e2 = 2; e2 <= largest_e; e2++) {
                /* Check gcd(e1, e2) == 1 */
                int a = e1, bb = e2;
                while (bb) { int t = bb; bb = a % bb; a = t; }
                if (a == 1) {
                    ans = (ans - 1 + M) % M;
                }
            }
        }
    }

    free(is_perfect_power);
    printf("%lld\n", ans % M);
    return 0;
}
